File: docs/vs-2015/snippets/visualbasic/VS_Snippets_VSSDK/firsttoolwin/vb/mytoolwindow.vb
Imports System
Imports System.Collections
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Windows
Imports System.Runtime.InteropServices
Imports Microsoft.VisualStudio.Shell.Interop
Imports Microsoft.VisualStudio.Shell
'<snippet21>
Imports System.ComponentModel.Design
'</snippet21>

''' <summary>
''' This class implements the tool window exposed by this package and hosts a user control.
'''
''' In Visual Studio tool windows are composed of a frame (implemented by the shell) and a pane, 
''' usually implemented by the package implementer.
'''
''' This class derives from the ToolWindowPane class provided from the MPF in order to use its 
''' implementation of the IVsUIElementPane interface.
''' </summary>
<Guid("efc69a2c-90f6-4dc9-8b74-4f2e20abcd43")> _
Public Class MyToolWindow
    Inherits ToolWindowPane

    '<snippet24>
    Dim control As MyControl
    Dim comboValue As String
    '</snippet24>

    ''' <summary>
    ''' Standard constructor for the tool window.
    ''' </summary>
    Public Sub New()
        MyBase.New(Nothing)
        ' Set the window title reading it from the resources.
        Me.Caption = Resources.ToolWindowTitle
        ' Set the image that will appear on the tab of the window frame
        ' when docked with an other window
        ' The resource ID correspond to the one defined in the resx file
        ' while the Index is the offset in the bitmap strip. Each image in
        ' the strip being 16x16.
        Me.BitmapResourceID = 301
        Me.BitmapIndex = 1

        'This is the user control hosted by the tool window; Note that, even if this class implements IDisposable,
        'we are not calling Dispose on this object. This is because ToolWindowPane calls Dispose on 
        'the object returned by the Content property.

        '<snippet22>
        ' Create the toolbar. 
        Me.ToolBar = New CommandID(GuidList.guidFirstToolWinCmdSet, PkgCmdIDList.cmdidWindowsMediaWin)
        Me.ToolBarLocation = CInt(VSTWT_LOCATION.VSTWT_TOP)

        ' Create the handlers for the toolbar commands. 
        Dim mcs As OleMenuCommandService = TryCast(GetService(GetType(IMenuCommandService)), OleMenuCommandService)
        If mcs IsNot Nothing Then
            Dim toolbarbtnCmdID As CommandID = New CommandID(GuidList.guidFirstToolWinCmdSet, PkgCmdIDList.cmdidMyCommand)
            Dim menuItem As MenuCommand = New MenuCommand(New EventHandler(AddressOf ButtonHandler), toolbarbtnCmdID)
            mcs.AddCommand(menuItem)

            ' Command for the combo itself 
            Dim menuMyDynamicComboCommandID As CommandID = New CommandID(GuidList.guidFirstToolWinCmdSet, CInt(PkgCmdIDList.cmdidMyCommand))
            Dim menuMyDynamicComboCommand As OleMenuCommand = New OleMenuCommand(New EventHandler(AddressOf ComboHandler), menuMyDynamicComboCommandID)
            mcs.AddCommand(menuMyDynamicComboCommand)

            ' Command for the combo's list 
            Dim comboListCmdID As CommandID = New CommandID(GuidList.guidFirstToolWinCmdSet, PkgCmdIDList.cmdidWindowsMediaWin)
            Dim comboMenuList As OleMenuCommand = New OleMenuCommand(New EventHandler(AddressOf ComboListHandler), comboListCmdID)
            mcs.AddCommand(comboMenuList)
        End If
        '</snippet22>

        Me.Content = New MyControl()
    End Sub

    '<snippet23>
    Private Sub ButtonHandler(ByVal sender As Object, ByVal arguments As EventArgs)
        If comboValue IsNot Nothing AndAlso comboValue.Trim().Length <> 0 Then
            LoadFile(comboValue)
        End If
    End Sub

    Private Sub ComboHandler(ByVal sender As Object, ByVal arguments As EventArgs)
        Dim eventArgs As OleMenuCmdEventArgs = TryCast(arguments, OleMenuCmdEventArgs)
        If eventArgs IsNot Nothing Then
            Dim output As IntPtr = eventArgs.OutValue
            Dim input As Object = eventArgs.InValue
            If input IsNot Nothing Then
                comboValue = input.ToString()
            ElseIf output <> IntPtr.Zero Then
                Marshal.GetNativeVariantForObject(comboValue, output)
            End If
        End If
    End Sub

    Public Sub LoadFile(ByVal comboValue As String)
        control.MediaPlayer.Source = New Uri(comboValue)
    End Sub

    Private Sub ComboListHandler(ByVal sender As Object, ByVal arguments As EventArgs)
    End Sub
    '</snippet23>

    '<snippet60>
    Public Overrides Sub OnToolWindowCreated()
        MyBase.OnToolWindowCreated()

        Dim handler As FirstToolWinPackage = CType( _
            Me.Package, FirstToolWinPackage)
        CType(Me.Frame, IVsWindowFrame).SetProperty( _
            CInt(__VSFPROPID.VSFPROPID_ViewHelper), handler)
    End Sub
    '</snippet60>

End Class

File: docs/vs-2015/snippets/csharp/VS_Snippets_VSSDK/firsttoolwin/cs/mytoolwindow.cs
using System;
using System.Collections;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Windows;
using System.Runtime.InteropServices;
using Microsoft.VisualStudio.Shell.Interop;
using Microsoft.VisualStudio.Shell;
//<snippet21>
using System.ComponentModel.Design;
//</snippet21>


namespace Company.FirstToolWin
{
    /// <summary>
    /// This class implements the tool window exposed by this package and hosts a user control.
    ///
    /// In Visual Studio tool windows are composed of a frame (implemented by the shell) and a pane, 
    /// usually implemented by the package implementer.
    ///
    /// This class derives from the ToolWindowPane class provided from the MPF in order to use its 
    /// implementation of the IVsUIElementPane interface.
    /// </summary>
    [Guid("26a8c5ea-ac69-4f4b-8408-22e16e0541a1")]
    public class MyToolWindow : ToolWindowPane
    {
        // This is the user control hosted by the tool window; it is exposed to the base class 
        // using the Content property. Note that, even if this class implements IDispose, we are
        // not calling Dispose on this object. This is because ToolWindowPane calls Dispose on 
        // the object returned by the Content property.
        //        private FrameworkElement control;
//<snippet24>
        private MyControl control;
        string comboValue = "";
//</snippet24>
        /// <summary>
        /// Standard constructor for the tool window.
        /// </summary>
        public MyToolWindow() :
            base(null)
        {
            // Set the window title reading it from the resources.
            this.Caption = Resources.ToolWindowTitle;
            // Set the image that will appear on the tab of the window frame
            // when docked with an other window
            // The resource ID correspond to the one defined in the resx file
            // while the Index is the offset in the bitmap strip. Each image in
            // the strip being 16x16.
            this.BitmapResourceID = 301;
            this.BitmapIndex = 1;
//<snippet22>
            // Create the toolbar.
            this.ToolBar = new CommandID(GuidList.guidFirstToolWinCmdSet,
                PkgCmdIDList.ToolbarID);
            this.ToolBarLocation = (int)VSTWT_LOCATION.VSTWT_TOP;

            // Create the handlers for the toolbar commands.
            var mcs = GetService(typeof(IMenuCommandService))
                as OleMenuCommandService;
            if (null != mcs)
            {
                var toolbarbtnCmdID = new CommandID(
                    GuidList.guidFirstToolWinCmdSet,
                    PkgCmdIDList.cmdidWindowsMediaOpen);
                var menuItem = new MenuCommand(new EventHandler(
                    ButtonHandler), toolbarbtnCmdID);
                mcs.AddCommand(menuItem);

                // Command for the combo itself
                var menuMyDynamicComboCommandID = new CommandID(
                    GuidList.guidFirstToolWinCmdSet,
                    (int)PkgCmdIDList.cmdidWindowsMediaFilename);
                var menuMyDynamicComboCommand = new OleMenuCommand(
                    new EventHandler(ComboHandler),
                    menuMyDynamicComboCommandID);
                mcs.AddCommand(menuMyDynamicComboCommand);

                // Command for the combo's list
                var comboListCmdID = new CommandID(
                    GuidList.guidFirstToolWinCmdSet,
                    PkgCmdIDList.cmdidWindowsMediaFilenameGetList);
                var comboMenuList = new OleMenuCommand(
                    new EventHandler(ComboListHandler), comboListCmdID);
                mcs.AddCommand(comboMenuList);
            } 
//</snippet22>
            control = new MyControl();
        }

        /// <summary>
        /// This property returns the control that should be hosted in the Tool Window.
        /// It can be either a FrameworkElement (for easy creation of toolwindows hosting WPF content), 
        /// or it can be an object implementing one of the IVsUIWPFElement or IVsUIWin32Element interfaces.
        /// </summary>
        override public object Content 
        {
            get
            {
                return this.control;
            }

            set
            {
                base.Content = value;
            }
        }
//<snippet23>
        private void ButtonHandler(object sender,
    EventArgs arguments)
        {
            if (comboValue != null && comboValue.Trim().Length != 0)
            {
                LoadFile(comboValue);
            }
        }

        private void ComboHandler(object sender, EventArgs arguments)
        {
            var eventArgs = arguments as OleMenuCmdEventArgs;
            if (eventArgs != null)
            {
                IntPtr output = eventArgs.OutValue;
                object input = eventArgs.InValue;
                if (input != null)
                {
                    comboValue = input.ToString();
                }
                else if (output != IntPtr.Zero)
                {
                    Marshal.GetNativeVariantForObject(comboValue,
                        output);
                }
            }
        }

        public void LoadFile(string comboValue)
        {
            control.MediaPlayer.Source = new System.Uri(comboValue);
        }

        private void ComboListHandler(object sender,
            EventArgs arguments)
        {
        }
//</snippet23>
//<snippet60>
        public override void OnToolWindowCreated()
        {
            base.OnToolWindowCreated();

            var handler = new ToolWindowEvents(
                (FirstToolWinPackage)this.Package);
            ((IVsWindowFrame)this.Frame).SetProperty(
                (int)__VSFPROPID.VSFPROPID_ViewHelper, handler);
        }
//</snippet60>
    }
}
